Calculate priority with number of active cores

tomtor opened the pull request:
See https://github.com/glinscott/fishtest/pull/458

@ppigazzini @vdv @xoto10 I have tested this locally and it is ready for some more testing and feedback.

This change allocates cores round-robin. It is not completely fair now, because all cores are treated equally, so some workers might get faster cores than others. This is easy to change to Mnps calculation, but because we show allocated cores on the main page this is both simple to code and least confusing for users.

Note that old existing tests will have 3000 as throughput, so until they are finished they will get a large amount of the cores. When merging we should change prio to 100 for old tests

ppigazzini commented:
@tomtor DEV updated, feel free to suggest how to test :)

ppigazzini commented:
DEV updated

tomtor commented:
@ppigazzini Thanks, I have allocated my 2 workers to DEV. Could you do the same?

All we can do is fiddle with some tests and look for strange events?

tomtor commented:
@ppigazzini Looks good, the run with 25% throughput get about 25% ...

ppigazzini commented:
@tomtor added 8 workers each w/ 3 cores.

tomtor commented:
@ppigazzini @vondele Looks good to me. It is obviously less accurate than the old method. A worker with lower priority could get eg a 32-core worker, when it is on average entitled to 3 cores, but it should all average out with a lot of workers over a longer time....

ppigazzini commented:
@tomtor I suddenly caught all the workers raising the throughput, cheating is so easy :)

tomtor commented:
I squashed....

vondele commented:
interesting... I wouldn't worry about the precise NPS of workers, or 'fractions of a worker'... this is just a detail.

I would do the adjustment for the TC internally... i.e. leave the throughput at 100% also for LTC on the submission website, but correct for TC internally, e.g. by giving all jobs a  (TC / 10s) boost. 

In fact, I would scale with
- sqrt(TC / 10s) 
- sqrt(cores)
- 0.5 for numgames type runs
- 0.5 for SPSA runs
- (5 + LLR)/5 for SPRT.

The rationale is to give more weight to SPRT tests (relative to numgames and spsa), and even more so if they are close to pass. I would sqrt-scale wrt to cores and TC, just because these jobs need more resources, but if we put too much weight on those, they will drain the framework for testing.

xoto10 commented:
There's a function somewhere that calculates the scaling factor for the
time control, you can just call that to allocate more codes to longer tc
tests. I would do that linearly, so ltc gets 6x cores of stc, at least for
now. And leave the other adjustments vdv suggests as well. They are
interesting to consider, but I think the initial implementation should be
simpler and closer to the current method. We can experiment with e.g. more
cores for jobs with higher LLR later.

My change is a lot less code :-)

tomtor commented:
@ppigazzini @vondele I implemented the suggestions of Joost and tested locally...

ppigazzini commented:
@tomtor DEV updated, I wrote some tests with different throughput and connected 10 workers.

vondele commented:
nice! The real load-balancing won't be visible unless there are many workers.

As soon as you feel the code is robust, this can be tested on the production server IMO, as an experiment. If this is not performing as expected it can be reverted or refined.

ppigazzini commented:
@tomtor @vondele started 30 mini workers.

tomtor commented:
> @tomtor @vondele started 30 mini workers.

@ppigazzini They all do 0.00 Mnps?

ppigazzini commented:
@tomtor 30 workers w/ 1 core put too much pressure on the VPS storage, I rebooted the server and started 10 workers w/ 3 cores.

tomtor commented:
@ppigazzini @vondele Note that you can see the `itp` (internal throughput) on the individual test pages from 
https://dfts-2.pigazzini.it/tests

ppigazzini commented:
@tomtor DEV updated

xoto10 commented:
> I would do the adjustment for the TC internally... i.e. leave the throughput at 100% also for LTC on the submission website, but correct for TC internally

@tomtor 
Not sure if you're doing this bit or not, I didn't see tc mentioned in your code but I only looked briefly. The function I was thinking of earlier is ` parse_tc() ` in views.py, very useful if you do want to automatically adjust for tc.

xoto10 commented:
> As soon as you feel the code is robust, this can be tested on the production server IMO, as an experiment. If this is not performing as expected it can be reverted or refined.

Agreed. I was hoping that would happen with my suggestion, but we're going for the rewrite approach instead. Whatever.

vondele commented:
My first impression is that this looks good ...

ppigazzini commented:
@tomtor some notes:
* I was able to set throughput <= 0: this makes sense?
* in test page perhaps I suggest to change "Adjust throughput:" in "Adjust Throughput%:"
* rescheduling an old test I got "Throughput%: 3000"

tomtor commented:
> > I would do the adjustment for the TC internally... i.e. leave the throughput at 100% also for LTC on the submission website, but correct for TC internally
> 
> @tomtor
> Not sure if you're doing this bit or not, I didn't see tc mentioned in your code but I only looked briefly. The function I was thinking of earlier is `parse_tc()` in views.py, very useful if you do want to automatically adjust for tc.

parse_tc does something else. It computes an expected game duration

xoto10 commented:
>  It computes an expected game duration

Yes, exactly. I thought that might be useful if you want to adjust core allocation based on expected game duration since it covers base time and increment.

vondele commented:
>     * I was able to set throughput <= 0: this makes sense?

No, we should keep this input >0 

>     * in test page perhaps I suggest to change "Adjust throughput:" in "Adjust Throughput%:"

agreed

>     * rescheduling an old test I got "Throughput%: 3000", a bit dangerous during the transition phase

I wonder if we should reject Throughput > e.g. 500. If needed we still have priority 1 for that.

ppigazzini commented:
@vondele @tomtor I'm wondering if we should have quantized values (e.g 10, 20,..., 100, default 60)

tomtor commented:
> No, we should keep this input >0

It is mapped to 1 when < 1, I will map > 500 to 500. I will change throughput of old tests by hand to 100.

I missed `Throughput%` for adjusting, changed that too and squashed.

@ppigazzini For me it's ok to merge and test in prod

tomtor commented:
> I'm wondering if we should have quantized values (e.g 10, 20,..., 100, default 60)

Why 60 and not 100? I think it's fine for now.

ppigazzini commented:
@tomtor just started commuting. At the moment there is a quantized set 1 - 500 with default 100, just wondering if we need so much headroom above the default.

tomtor commented:
@ppigazzini The old implementation had no checking or limits on throughput at all. Users can also raise priority which has always allocated all cores. So the 1-500 range is now more limiting than the original code.

ppigazzini commented:
@tomtor @vondele Before merging and updating the PROD server IMO we should write a post on fishcooking to inform about the change.

tomtor commented:
> @vondele @tomtor I'm wondering if we should have quantized values (e.g 10, 20,..., 100, default 60)

@ppigazzini @vondele I do agree that this could be an improvement when implemented with a popup selector with values:

10%, 25%, 50%, 100% (default),  200% (in red)

tomtor commented:
I added the selector for NEW tests. Updating is still with typed input so a range of 1-500 remains possible. Best of both worlds.

tomtor commented:
> @tomtor @vondele Before merging and updating the PROD server IMO we should write a post on fishcooking to inform about the change.

I will write one...

ppigazzini commented:
Nice patch @tomtor , thank you to @xoto10 to starting the improvement.

tomtor commented:
@ppigazzini Merged, restarted and changed old runs to prio 100 (and 50/25)

ppigazzini commented:
@tomtor I'm viewing on the web. Finally the throughput value makes sense.

xoto10 commented:
Nice work @tomtor !

xoto10 commented:
Looks ok as far as I can tell so far. :-)
I would allocate the full 6x cores for LTC, but only half of what you're allocating in the threaded cases, but no doubt there will be many different opinions on that kind of detail, we can canvas opinion on that once it's been running for a while ...

vondele commented:
Yes, it seems to be working rather well:

![image](https://user-images.githubusercontent.com/4202567/70787574-439b4700-1d8f-11ea-853a-d01b33de1c6e.png)

rather balanced distribution of cores to tasks, no pending tasks, more cores in the more promising tests.

tomtor commented:
One thing that could be changed now is creating smaller batches (eg 100) for ltc/ multi core because they can occupy workers a long time...

vondele commented:
@tomtor that would be really a nice addition as well. Certainly worker side, since we cache the SF builds, I think the overhead of having smaller batches must be relatively small.

You'll be better aware of the possible tricky issues server-side (performance of the server, or 'accounting' of node hours to users donating time).

xoto10 commented:
Odd distribution of workers just now:

![image](https://user-images.githubusercontent.com/23479932/70798439-df807f00-1d9e-11ea-8951-9744512b9116.png)

Two long-running tunes with itp 12 and 30, seems odd one would get almost all the workers like this.

vondele commented:
Actually, right now looks a bit surprising:

![image](https://user-images.githubusercontent.com/4202567/70798405-2b372680-1da7-11ea-8a09-26002ff7c48e.png)

I would expect to see the two tunes have roughly the same amount of cores (corrected by the user-set throughput), as they were assigned all available cores. The internal throughput seems right, but the number of cores assigned is vastly different, as if all cores were assigned to the test with the highest internal throughput.

tomtor commented:
@vondele But `ps_pawnscoring6` has prio `-1` so it SHOULD get no new workers !

ppigazzini commented:
@tomtor enhancements:
* expose the priority/throughput values, e.g. `p: 0 t: 100 c: 993`
* compute p-values and residuals to get some real time indexes about the fair cores allocation

This should help to keep a low fishtest's developers stress level.

Just kidding :)

vondele commented:
@tomtor thanks... I overlooked the priority -1 ... so more evidence it is working well :-)
